src/librustc_mir/hair/cx/mod.rs

```diff
@@ -169,38 +156,46 @@ impl<'a, 'gcx, 'tcx> Cx<'a, 'gcx, 'tcx> {
         };
 
         let clamp = |n| {
-            let size = self.integer_bit_width(ty);
-            trace!("clamp {} with size {} and amt {}", n, size, 128 - size);
-            let amt = 128 - size;
+            let gcx = self.tcx.global_tcx();
+            let param_ty = self.param_env.and(gcx.lift(&ty).unwrap());
+            let bit_width = gcx.layout_of(param_ty).unwrap().size.bits();
+            trace!("clamp {} with size {} and amt {}", n, bit_width, 128 - bit_width);
+            let amt = 128 - bit_width;
             let result = (n << amt) >> amt;
```

Comment:
Can't this use some `truncate` function from `miri`?

Reply:
won't fix in this PR, there's an issue open for cleaning up this kind of code: https://github.com/rust-lang/rust/issues/49937